kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -256,67 +239,76 @@ protected TransactionImpl buildTransactionInternal(
-    // need to execute our protocol & metadata validation + update logic and possibly get an updated
-    // protocol and/or metadata
-    Metadata baseMetadata;
-    Protocol baseProtocol;
-    if (isCreateOrReplace) {
-      // For a new table definition start with an empty initial metadata
-      baseMetadata = getInitialMetadata();
-      // In the case of Replace table there are a few delta-specific properties we want to preserve
-      if (latestSnapshot.isPresent()) { // replace = isCreateOrReplace && latestSnapshot.isPresent
-        Map<String, String> propertiesToPreserve =
-            latestSnapshot.get().getMetadata().getConfiguration().entrySet().stream()
-                .filter(
-                    e ->
-                        ReplaceTableTransactionBuilderImpl.TABLE_PROPERTY_KEYS_TO_PRESERVE.contains(
-                            e.getKey()))
-                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        baseMetadata = baseMetadata.withMergedConfiguration(propertiesToPreserve);
-      }
-    } else {
-      // Otherwise, use the existing table metadata
-      baseMetadata = latestSnapshot.get().getMetadata();
+    // Instead of special casing enabling clustering or domain metadata, we should just add them
+    // to the table properties which we already handle.
+    Map<String, String> tablePropertiesWithFeatureEnablement =
+        new HashMap<>(tableProperties.orElse(emptyMap()));
+    boolean domainMetadataEnabled =
+        !isCreateOrReplace
+            && latestSnapshot
+                .get()
+                .getProtocol()
+                .supportsFeature(TableFeatures.DOMAIN_METADATA_W_FEATURE);
+    if (needDomainMetadataSupport && !domainMetadataEnabled) {
+      tablePropertiesWithFeatureEnablement.put(
+          TableFeatures.SET_TABLE_FEATURE_SUPPORTED_PREFIX
+              + TableFeatures.DOMAIN_METADATA_W_FEATURE.featureName(),
+          "supported");
     }
-    if (latestSnapshot.isPresent()) {
-      // If latestSnapshot is present it is either a write to an existing table or a replace table.
-      // In both cases we want to start with the prior table protocol to ensure we never downgrade
-      // protocols.
-      baseProtocol = latestSnapshot.get().getProtocol();
-    } else {
-      // Otherwise, start with initial protocol for a new table
-      baseProtocol = getInitialProtocol();
+    boolean clusteringEnabled =
+        !isCreateOrReplace
+            && latestSnapshot
+                .get()
+                .getProtocol()
+                .supportsFeature(TableFeatures.CLUSTERING_W_FEATURE);
+    if (inputLogicalClusteringColumns.isPresent() && !clusteringEnabled) {
+      tablePropertiesWithFeatureEnablement.put(
+          TableFeatures.SET_TABLE_FEATURE_SUPPORTED_PREFIX
+              + TableFeatures.CLUSTERING_W_FEATURE.featureName(),
+          "supported");
+    }
```

Comment:
TODO: move this auto-enablement to inside of `TransactionMetadataFactory` -- the only one we should have to special case here is `withDomainMetadataSupported`